bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java

```diff
@@ -0,0 +1,80 @@
+        if (!typeParameterMap.hasTypeParameters() || typeParameters.isEmpty()) {
+            return typeData;
+        }
+
+        Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap = typeParameterMap.getPropertyToClassParamIndexMap();
+        Either<Integer, TypeParameterMap> classTypeParamRepresentsWholeField = propertyToClassParamIndexMap.get(-1);
+        if (classTypeParamRepresentsWholeField != null) {
+            Integer index = classTypeParamRepresentsWholeField.map(i -> i, e -> {
+                throw new IllegalStateException("Invalid state, the whole class cannot be represented by a subtype.");
+            });
+            return (TypeData<V>) typeParameters.get(index);
+        } else {
+            return getTypeData(typeData, typeParameters, propertyToClassParamIndexMap);
+        }
+    }
+
+    private static <V> TypeData<V> getTypeData(final TypeData<V> typeData, final List<TypeData<?>> specializedTypeParameters,
+                                               final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap) {
+        List<TypeData<?>> subTypeParameters = new ArrayList<>(typeData.getTypeParameters());
+        for (int i = 0; i < typeData.getTypeParameters().size(); i++) {
+            subTypeParameters.set(i, getTypeData(subTypeParameters.get(i), specializedTypeParameters, propertyToClassParamIndexMap, i));
+        }
+        return TypeData.builder(typeData.getType()).addTypeParameters(subTypeParameters).build();
+    }
+
+    private static TypeData<?> getTypeData(final TypeData<?> typeData, final List<TypeData<?>> specializedTypeParameters,
+                                           final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap,
+                                           final int index) {
+        if (!propertyToClassParamIndexMap.containsKey(index)) {
+            return typeData;
+        }
+        return propertyToClassParamIndexMap.get(index).map(l -> {
+                    if (typeData.getTypeParameters().isEmpty()) {
+                        // Represents the whole typeData
+                        return specializedTypeParameters.get(l);
+                    } else {
+                        // Represents a single nested type parameter within this typeData
+                        TypeData.Builder<?> builder = TypeData.builder(typeData.getType());
+                        List<TypeData<?>> typeParameters = new ArrayList<>(typeData.getTypeParameters());
+                        typeParameters.set(index, specializedTypeParameters.get(l));
+                        builder.addTypeParameters(typeParameters);
+                        return builder.build();
+                    }
+                },
+                r -> {
+                    // Represents a child type parameter of this typeData
+                    return getTypeData(typeData, specializedTypeParameters, r.getPropertyToClassParamIndexMap());
```

Comment:
Nothing needs specializing for this type, but one of this types type parameters needs specializing, so we recurse down and specialize there.